x/service/keeper/msg_server_add_service.go

```diff
@@ -23,14 +24,57 @@ func (k msgServer) AddService(
 	logger := k.Logger(ctx).With("method", "AddService")
 	logger.Info(fmt.Sprintf("About to add a new service with msg: %v", msg))
 
+	// Validate the message.
 	if err := msg.ValidateBasic(); err != nil {
 		logger.Error(fmt.Sprintf("Adding service failed basic validation: %v", err))
 		return nil, err
 	}
 
+	// Check if the service already exists or not.
 	if _, found := k.GetService(ctx, msg.Service.Id); found {
 		logger.Error(fmt.Sprintf("Service already exists: %v", msg.Service))
-		return nil, types.ErrServiceAlreadyExists
+		return nil, types.ErrServiceAlreadyExists.Wrapf(
+			"duplicate ID: %s", msg.Service.Id,
+		)
+	}
+
+	// Retrieve the address of the actor adding the service.
+	accAddr, err := sdk.AccAddressFromBech32(msg.Address)
+	if err != nil {
+		logger.Error(fmt.Sprintf("could not parse address %s", msg.Address))
+		return nil, types.ErrServiceInvalidAddress.Wrapf(
+			"%s is not in Bech32 format", msg.Address,
+		)
+	}
+
+	// Check the actor has sufficient funds to pay for the add service fee.
+	accCoins := k.bankKeeper.SpendableCoins(ctx, accAddr)
+	if accCoins.Len() == 0 {
+		logger.Error(fmt.Sprintf("%s doesn't have any funds to add service: %v", msg.Address, err))
+		return nil, types.ErrServiceNotEnoughFunds.Wrapf(
+			"account has no spendable coins",
+		)
+	}
+
+	// Check the balance of upokt is enough to cover the AddServiceFee.
+	accBalance := accCoins.AmountOf("upokt")
+	if accBalance.LTE(sdk.NewIntFromUint64(k.GetParams(ctx).AddServiceFee)) {
```

Comment:
NIT: Can you move `sdk.NewIntFromUint64(k.GetParams(ctx).AddServiceFee))` into a local var for readability?